Don't clear managedFields on live object

alexzielenski opened the pull request:
/hold
/kind bug

Fixes a bug wherein the managedFields of the `liveObject` are altered with calls to Update/Apply. This is problematic because there may be webhooks which expect to receive change notifications with a copy of the old object and the new object. Changes to the live object can cause the notification to be inaccurate.

This patch adds a test to make sure that the liveObject instance returned by a call to `FieldManager.Apply` has its `managedFields` metadata intact after another call to `FieldManager.Apply`.

Also adds two tests to check that managedFields is unaltered (order, timestamp) when a configuration is applied which should lead to a noop. (For both Apply & Update code paths).

```release-note
Fixed a bug which could cause webhooks to have an incorrect copy of the old object after an Apply or Update
```

apelisse commented:
/assign

apelisse commented:
Thank you Alex, that's great!

A quick note on an inaccuracy in the description:
> This is problematic because there are instances where clients of this code may have direct references to liveObject and expect it not to change.

Clients don't have direct references, but there can be webhooks (third-party processes that can see the modifications happening on the cluster) that receives a copy of the live object and the new object. Because of this bug, they receive an incorrect version of the live object.

apelisse commented:
/ok-to-test

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go:
```diff
@@ -1291,3 +1291,273 @@ func TestUpdateViaSubresources(t *testing.T) {
+}
+
+func TestLiveObjectManagedFieldsNotRemoved(t *testing.T) {
+	// This test makes sure that the liveObject during a patch does not mutate
```
Nit, we typically put this comment before the function rather than at the beginning of the function

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go:
```diff
@@ -1291,3 +1291,273 @@ func TestUpdateViaSubresources(t *testing.T) {
+		t.Fatalf("failed to apply object: %v", err)
+	}
+
+	originalLiveObj := f.liveObj
```
I find using `originalLiveObj` rather than `f.liveObj` a little confusing, wouldn't it work the same if you were using `f.liveObj` below (both before and after apply)?

alexzielenski replied:
The call to Apply is liable to re-assign `f.liveObj` https://github.com/kubernetes/kubernetes/blob/ef6d440db36348563f2f68e087a83ccd32f3671a/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go#L169

apelisse reviewed (commented):
Looks good to me, I'm happy we're able to test with such minimal setup.

@liggitt WDYT?

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go:
```diff
@@ -1291,3 +1291,273 @@ func TestUpdateViaSubresources(t *testing.T) {
+		"spec": {
+			"replicas": 101,
+		}
+	}`), &appliedObj.Object); err != nil {
```
This object isn't used in that test.

apelisse commented:
Something definitely broke because of this change:
```
=== FAIL: vendor/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager TestNoOpChanges (3.20s)
    fieldmanager_test.go:809: failed to re-apply object: metadata.managedFields must be nil
```

liggitt commented:
> === FAIL: vendor/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager TestNoOpChanges (3.20s)

the test had a bug in it, see fixup in https://github.com/kubernetes/kubernetes/pull/105852/files#diff-16ff3ca9b9a959368bf1ed86bbe146fda7df824426129eceea577a4b4b032bc6

liggitt commented:
We want to make sure what we return from `Apply` does not have managedFields set. Since we're no longer clearing them from `liveObj` up front, when we end up returning `liveObj` from `Apply`, I think we need to copy/clear managedFields at that point.

See https://github.com/kubernetes/kubernetes/pull/105852/files#diff-ec1c16b4530fabf58319ab8770e106e8a2b6277ecb85f0031d9ac9d0d18adb4d (needs to be exercised with a test)

apelisse commented:
> We want to make sure what we return from `Apply` does not have managedFields set. Since we're no longer clearing them from `liveObj` up front, when we end up returning `liveObj` from `Apply`, I think we need to copy/clear managedFields at that point.

Ah yes, good point! Thanks!

alexzielenski commented:
> We want to make sure what we return from `Apply` does not have managedFields set. Since we're no longer clearing them from `liveObj` up front, when we end up returning `liveObj` from `Apply`, I think we need to copy/clear managedFields at that point.
> 
> See https://github.com/kubernetes/kubernetes/pull/105852/files#diff-ec1c16b4530fabf58319ab8770e106e8a2b6277ecb85f0031d9ac9d0d18adb4d (needs to be exercised with a test)

Does this bit of the StripMetaManager part of the pipeline not already strip out the metadata fields of the returned object or am I misinterpreting what it is doing?

https://github.com/kubernetes/kubernetes/blob/ef6d440db36348563f2f68e087a83ccd32f3671a/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/stripmeta.go#L72

liggitt commented:
> Does this bit of the StripMetaManager part of the pipeline not already strip out the metadata fields of the returned object or am I misinterpreting what it is doing?

no, that doesn't modify `newObj`

alexzielenski committed:
guard against modification of test inputs in TestNoOpChanges

alexzielenski committed:
remove mutation of liveObj in fieldManager.Apply

also document manager interface

alexzielenski commented:
/retest

fedebongio commented:
/triage accepted

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager_test.go:
```diff
@@ -780,14 +780,14 @@ func TestNoOpChanges(t *testing.T) {
 		t.Fatalf("error decoding YAML: %v", err)
 	}
-	if err := f.Apply(obj, "fieldmanager_test_apply", false); err != nil {
+	if err := f.Apply(obj.DeepCopyObject(), "fieldmanager_test_apply", false); err != nil {
```
Yes right, that test was broken, thanks!

alexzielenski commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/managedfieldsupdater.go:
```diff
@@ -80,7 +81,8 @@ func (f *managedFieldsUpdater) Apply(liveObj, appliedObj runtime.Object, managed
 	if object == nil {
-		object = liveObj
+		object = liveObj.DeepCopyObject()
+		internal.RemoveObjectManagedFields(object)
```
This line is exercised here: https://github.com/kubernetes/kubernetes/pull/106195/files#diff-16ff3ca9b9a959368bf1ed86bbe146fda7df824426129eceea577a4b4b032bc6R1374

I just manually verified that it is hit. I'm wondering if you know of a better way to make the connection between the test and that line's coverage explicit? Currently the test only checks that `liveObj` is unchanged.

apelisse replied:
You maybe need to write a more target unit-test, probably in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/managedfieldsupdater_test.go

liggitt replied:
we want a test that not only hits this line, but if possible, would fail without the deepcopy and dropping managed fields

apelisse replied:
Yeah, but this line is basically completely pointless because `FieldManager` always replaces it anyway

apelisse replied:
If we want to test that it's being reset, we need to test without the `FieldManager`.

alexzielenski replied:
I've pushed my attempt to address the above feedback in 99fd43965a37a3552acf67554790d386a4d7afbd please take a look. This commit has a test which adds a node at the end of the manager pipeline, but before `FieldManager`, so that we can directly test if the `liveObj` has been copied, and its managedFields removed.

alexzielenski committed:
verify liveObj not returned by manager pipeline

addresses feedback by adding a test to make sure that the manager pipeline will not return the same instance used as input, and that the output does not input managedFields

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/managedfieldsupdater_test.go:
```diff
@@ -64,3 +67,87 @@ func TestNoManagedFieldsUpdateDoesntUpdateTime(t *testing.T) {
+
+	updater := NewManagedFieldsUpdater(NoopManager{})
+
+
+	// Take a copy before applying so we can check that live has not been changed
+	originalLiveObject := obj.DeepCopyObject()
+
```
```suggestion
```

I don't know yet if this is necessary or not, but since that's part of the contract it probably doesn't hurt.

alexzielenski replied:
Yeah this was just being diligent to enforce the contract. I suppose it is out of scope for this unit test. I removed the lines related to that.

apelisse reviewed (commented):
/lgtm
/approve

apelisse commented:
gofmt is a legit error, I'll re-lgtm once fixed, thanks!

apelisse commented:
/lgtm

alexzielenski commented:
/hold cancel

alexzielenski commented:
Seems like this test is not deterministic?

apelisse commented:
/retest

apelisse commented:
Yeah, if you have patience, you can find out who wrote/modified that test and open a ticket for the flake. cc @liggitt

liggitt commented:
69b9f9b1f0293fd56917d310b79b2e4be3672d9c

liggitt commented:
cc @jingxu97 on flake in Test_Run_OneVolumeDetachFailNodeWithReadWriteOnce
